### ISSUE TITLE
Time to live (TTL)

### ISSUE THREAD

**Issue by tekumara** [slatedb/slatedb#218]
**Is your feature request related to a problem? Please describe.**

To manage GDPR obligations we use the time to live (TTL) feature of DynamoDB. Having this feature would support a migration to SlateDB.

**Describe the solution you'd like**

Similar to DynamoDB and RocksDB I'd like to be able to specify a timestamp for each key. It would be used to expire the key/value for a specified TTL.

**Describe alternatives you've considered**

Our application could manage deletion on a schedule instead.

**Additional context**

https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/TTL.html

https://github.com/facebook/rocksdb/wiki/Time-to-Live

**Comment by criccomini** [comment]
Ah, yes. This is a reasonable feature to add. I think this requires a few things:

1. Adding a TTL option to the `WriteOptions` struct
2. Persisting the TTL (when set) for a key-value pair (updating BlockBuilder)
3. Updating the compactor to drop ignore any k-v pairs that have expired
4. Add a compactor config setting to force compaction on any SR older than some age

I'm thinking we need (4) so we can guarantee that the compactor will have the opportunity to drop TTL'd k-v pairs on old SRs that don't get touched that often.

**Comment by criccomini** [comment]
Looking at [RocksDB's TTL implementation](https://github.com/facebook/rocksdb/wiki/Time-to-Live). Looks like they suffix the TTL seconds to the value field. We can probably extend what we're doing in `ValueDeletable`. We currently have a per-value `len` u32 in the `BlockBuilder` that we overload to double as a delete flag (when it's sent to u32::MAX). Maybe it's time to get rid of this and add a second u32 bit flag header instead. We could have a bit signifying deletion, some number of bits for TTL timestamp, and perhaps other things as well.

We might want to think through whether it makes to house the TTLs elsewhere, though (in the block or SST-level). It depends a bit on the use case, I think. If we have relatively few values with TTLs, we might want to just house them in the SST or block to eschew added per-k/v overhead.

**Comment by criccomini** [comment]
SGTM! I think this feature needs a design doc.

**Comment by agavra** [comment]
Happy to take this one on 🫡 will try to come up with a doc in the next week or so

**Comment by criccomini** [comment]
@agavra Awesome! Thanks 😃 assigning to you.

**Comment by agavra** [comment]
Just adding a note for myself - looks like RocksDB uses [Compaction Filters](https://github.com/facebook/rocksdb/wiki/Compaction-Filter) to implement TTL: https://github.com/facebook/rocksdb/blob/6549b117148f01a5c35287d5de9754183dd3b781/utilities/ttl/db_ttl_impl.h#L149C7-L149C26

Some tradeoffs here. Allegedly:
> Using compaction filters has some serious limitations, however. For example, improperly using compaction filters can break some basic data consistency guarantees, and snapshot reads may no longer be repeatable (if the data is modified between reads). Hence, compaction filters are easier to use when consistency is not required. We consider the question of how to implement compaction filters more safely an open problem. Another limitation of compaction filters is that they do not allow multiple KV pairs to be atomically dropped or modified; e.g., it is not possible to atomically drop a KV-pair and the corresponding data in a secondary index. Despite these limitations, compaction filters are widely used, which in our view demonstrates that even imperfect solutions can go a long way if the application can find appropriate workarounds.

Source: https://dl.acm.org/doi/pdf/10.1145/3483840

Trying to find if there are other valid strategies for TTL on non-timeseries data.

**Comment by criccomini** [comment]
Do these compaction filter limitations apply to filters that implement TTL? Or is the statement more about filters that have arbitrary logic and might do odd things. I'm having a bit of trouble understanding how the TTL could result in consistency issues. Is it because the compactor might compact out a higher-level key whose TTL has expired, but hte same key might exist in a lower level that hasn't yet been compacted out (and whose TTL isn't set or hasn't yet expired)?

**Comment by agavra** [comment]
I was trying to work that through as well, not totally sure I see how it can result in consistency issues either. I was trying to enumerate some possible situations, especially with out-of-order updates that change TTL for one key but I haven't figured out anything that would cause an issue yet.

> Is it because the compactor might compact out a higher-level key whose TTL has expired, but hte same key might exist in a lower level that hasn't yet been compacted out (and whose TTL isn't set or hasn't yet expired)?

I looked through how it works and this should never happen either - if you "compact it out" you actually insert a tombstone, so the tombstone wouldn't be collected until it's all the way in the lowest level of the tree.

**Comment by agavra** [comment]
OK, if I'm reading the details about snapshots in RocksDB correctly it seems that a snapshot doesn't actually prevent things from being compacted but instead the compactor is "aware" of the snapshot data and ensures that it doesn't get removed: 

> Both flush and compaction use CompactionIterator to process key-value pairs and determine if each key-value pair should be dropped or output to resulting file. CompactionIterator is aware of all the snapshots and ensures that the data visible to each snapshot is preserved. [...] During flush/compaction, we have to scan the snapshot list when we need to find out the earliest snapshot to which a key is visible.
> 
> Source: https://github.com/facebook/rocksdb/wiki/Snapshot 

Perhaps this means that it's possible for a compaction filter to run and filter out TTL'd elements even if they're referenced by a snapshot? That would just seem like a bug to me. It seems like that is indeed the case:

> Before release 6.0, if there is a snapshot taken later than the key/value pair, RocksDB always try to prevent the key/value pair from being filtered by compaction filter so that users can preserve the same view from a snapshot, unless the compaction filter returns IgnoreSnapshots() = true. However, this feature is deleted since 6.0, after realized that the feature has a bug which can't be easily fixed. Since release 6.0, with compaction filter enabled, RocksDB always invoke filtering for any key, even if it knows it will make a snapshot not repeatable.
> 
> Source: https://github.com/facebook/rocksdb/wiki/Compaction-Filter

**Comment by agavra** [comment]
Did some more digging and found the bug they're referring to: 

> Disallow CompactionFilter::IgnoreSnapshots() = false, because it is not very useful and the behavior is confusing. The filter will filter everything if there is no snapshot declared by the time the compaction starts. However, users can define a snapshot after the compaction starts and before it finishes and this new snapshot won't be repeatable, because after the compaction finishes, some keys may be dropped.
>
> Source: https://github.com/facebook/rocksdb/releases/tag/v6.0.1

**Comment by agavra** [comment]
OK reading back through the SlateDB snapshot design and the [thread on Discord](https://discord.com/channels/1232385660460204122/1232385660946616433/1283136584719274004) I think we won't have this problem. The snapshot references a specific manifest, which has a specific set of SSTs. If compaction runs on those SSTs, anything referencing the old SSTs would still be consistent if the read was repeated on the same snapshot.

This is only a problem if snapshots are implemented with sequence numbers, the way RocksDB does it.

I'm feeling like a compaction filter for TTL makes a lot of sense with our current approach for snapshotting.

**Comment by agavra** [comment]
Dropping some more research notes for myself regarding where to encode timestamps and how to generate them. I think it would be good to make sure we support first-class user-defined timestamps from day 0 instead of requiring them to be part of the key or value serde.

> Applications can work around these limitations by encoding timestamps within the key or
within the value. However, they will experience performance degradations in either case. Encoding within the key sacrifices performance for point-lookups, while encoding within the value sacrifices performance for out-of-order writes to the same key and complicates the reading of old versions of keys. We believe adding support for application-specified timestamps would better address these limitations, where the application can tag its data with timestamps that can be understood globally, and do so outside the key or value
> 
> We have added basic support for application-specified timestamps [27](https://github.com/facebook/rocksdb/wiki/User-defined-Timestamp) and evaluated this approach with DB-Bench. [...] Relative to the baseline where the application encodes a timestamp as part of the key (transparent to RocksDB), the application-specified timestamp API can lead to a 1.2× or better throughput gain. The improvements arise from treating the timestamp as metadata separate from the key, because then point lookups can be used instead of iterators to get the newest value for a key, and Bloom filters may identify SSTables not containing that key. Additionally, the timestamp range covered by an SSTable can be stored in its properties, which can be leveraged during a search to exclude SSTables that would only contain stale values.

**Comment by vigneshc** [comment]
I assume the TTL gets reset when a key is re-inserted.

**Comment by vigneshc** [comment]
> Dropping some more research notes for myself regarding where to encode timestamps and how to generate them. I think it would be good to make sure we support first-class user-defined timestamps from day 0 instead of requiring them to be part of the key or value serde.

Would we treat timestamp as local clock timestamp or would it be user provided, similar to Kafka's `CreateTime`? I think `CreateTime` like semantics has its use cases. I see above is hinting towards user provided timestamp.

**Comment by vigneshc** [comment]
In addition to TTL implemented using compaction filter, it would be good to also implement it using a metadata only operation. If SsTableInfo had information about included timestamp range, entire SST can be dropped while enforcing TTL.

**Comment by KaguraMilet** [comment]
By the way, the folks in Boston have some good research on TTL.
1. [Lethe](https://disc-projects.bu.edu/lethe/)
2. [Acheron: Persisting Tombstones in LSM Engines](https://dl.acm.org/doi/10.1145/3555041.3589719)

**Comment by agavra** [comment]
> I assume the TTL gets reset when a key is re-inserted.

I think there's two types of TTL:
- implicit, session/table-wide (i.e. applies to every insert)
- set per-key

For the former, I think it makes sense to reset when a key is re-inserted. For the latter (assuming the former is not configured) it will be set to whatever the new insertion has set as its TTL. I'll write this all up in the design doc I'm working on.

> Would we treat timestamp as local clock timestamp or would it be user provided, similar to Kafka's CreateTime? I think CreateTime like semantics has its use cases. I see above is hinting towards user provided timestamp.

Exactly. I was thinking you could either configure a clock for your session and/or manually configure timestamps for individual operations. That would look something like:

```rs
let opts = DbOptions {..., clock: SystemTime::now}; // this would be the default (forget syntax)
...
db.put(key, value, timestamp); // the third is optional and defaults to the clock time
```

> In addition to TTL implemented using compaction filter described in https://github.com/slatedb/slatedb/issues/225 , it would be good to also implement it using a metadata only operation. If SsTableInfo had information about included timestamp range, entire SST can be dropped while enforcing TTL.

➕ yup totally agree! this would be similar to TWCS in cassandra: https://cassandra.apache.org/doc/stable/cassandra/operating/compaction/twcs.html

**Comment by agavra** [comment]
I wanted to record different use cases for TTL/timestamps in stream processing since there was some discussion around this on discord (these are drawn from real conversations - see https://issues.apache.org/jira/browse/KAFKA-4212):

There are basically two types of use cases for TTL:

1. Space Saving. This is the most common request, where the TTL is set to something large enough that by the time it triggers in users are confident they don't need that data anymore. Deduplication is a common use case for this (you only need to store everything you've seen for X days).
2. Correctness. This is less common, but an example of a use case here is maintaining a table of access tokens that have a TTL (and can be refreshed). In streams, you might want to join a stream of data with this table of access tokens to verify that the stream refers to valid access tokens. Consider this example:

```
(A) insert access_token1 ts=0 expire_at=100
(B) update access_token1 ts=200 expire_at=300
(C) delete access_token1 ts=250
```

There are two interesting "time" situations:

1. Read Interleaving (Time-Based MVCC): I may want to read the value at `ts=150` because the event on the stream arrived after I updated the table. This should return nothing for `access_token1` but if I issued it at table `ts=220` I would get a valid token. This is what RocksDB user timestamps addresses (my original suggestion didn't really cover this). KStreams partially solves for this: https://cwiki.apache.org/confluence/display/KAFKA/KIP-889%3A+Versioned+State+Stores

2. Out of Order Writes: imagine an interleaving where the writes come in as A -> C -> B. If A read comes in at `ts=280` it would find a valid `access_token1` because B "revived" the deleted row. 

Obviously if you can mix both of the problems above, and then it gets really complicated... you may want to read an out-of-order write because of an interleaved read.

tl;dr, I updated the proposal in #228 to just focus on the more common use case, which is saving space with TTL.

**Comment by agavra** [comment]
Closing issue since all work is complete 🥳
